clang/lib/CIR/CodeGen/CIRGenStmt.cpp

```diff
@@ -258,9 +255,12 @@ mlir::LogicalResult CIRGenFunction::buildSimpleStmt(const Stmt *S,
   case Stmt::DeclStmtClass:
     return buildDeclStmt(cast<DeclStmt>(*S));
   case Stmt::CompoundStmtClass:
-    return useCurrentScope
-               ? buildCompoundStmtWithoutScope(cast<CompoundStmt>(*S))
-               : buildCompoundStmt(cast<CompoundStmt>(*S));
+    // FIXME(cir): This is a form of early optimization. Perhaps we should
+    // always emit a `cir.scope` if one exists in the source code, then
+    // canonicalize it away with the merge-cleanup pass.
```

Comment:
This is more about having options: sometimes is more convenient to create a scope outside `buildCompoundStmt`, it's not really about optimization, which we don't do: https://godbolt.org/z/4odeYds5j

Reply:
I'll remove it, but I'm not sure if we are on the same page here.

My main issue with this conditional is that it omits the `cir.scope` op in certain scenarios. A `cir.if`, for example, will not have an explicitly defined `cir.scope`  operation in its regions.

In theory, this diverges from the original codegen to early-optimize (or maybe early-canonicalize) the code, which are two things we wish to avoid by principle. In practice, I do think it makes the code cleaner.

Reply:
> My main issue with this conditional is that it omits the `cir.scope` op in certain scenarios. A `cir.if`, for example, will not have an explicitly defined `cir.scope` operation in its regions.

This is by design, it's not an early optimization - see how the lifetime checker deals with if regions. Whether we should keep it looking into the future, it's another story, another PR. The initial idea is that it was redundant to add a scope op that basically wraps all of the region, so the regions for a if (and some other ops) imply scope like behavior. I agree how it could be confusing though (you just seemed to be caught here too) and I just noticed I didn't document that behavior. Alternatives approach could be:

1. Add an extra class member to every `CIR_Op`, something like `regionsHaveScopeLikeBehavior()`. Have that be `false` by default and turn that true for ops that embed the semantics.
2. Explicitly add a scope to every region that makes sense, I kinda dislike the verbosity though:
```
cir.if(...) {
 cir.scope {
 ...
 }
} else ...
``` 

If you wanna document the implying scope behavior, please add those FIXME's where it's more clear what you want to do in the future.

Reply:
Created an issue for this in #335

Reply:
> see how the lifetime checker deals with if regions.

Seems like we check each operation individually. A suggestion here is to use traits. In fact, I believe the `AutomaticAllocationScope` is already properly used by operations that have an implied scope.

> it was redundant to add a scope op that basically wraps all of the region, so the regions for a if (and some other ops) imply scope-like behavior.

It is indeed redundant. In practice, I do think we are better without it, but I also believe there might be a better name for `cir.scope`. Something like `cir.compound`, for example.